specs/projectreuniondeployment/projectreuniondeployment.md

```diff
@@ -0,0 +1,138 @@
+Reunion they require is already available, so this feature will provide an API to check that
+availability and provide links for resolving any deployment issues. This API could be expanded
+in the future to wrap deployment behavior for software capable of deploying packages and/or OS
+updates to reduce the end-user friction and minimize potential for developers to lose customers
+due to dropoff during the acquistion of Project Reunion.
+
+# Description
+
+When a Project Reunion-consuming package or app is deployed, it is expected to only have the App
+package or installer itself, and the Project Reunion main and framework packages may not be
+present, or may not be at the desired version. This API enables a developer to detect when the
+required Project Reunion main and/or framework packages are present, and to direct the user to the
+appropriate resources to acquire or update Project Reunion if it is absent or requires updates.
+
+Since this API is part of the Project Reunion framework, the framework itself must be present.
+For MSIX packages, the framework must be listed as a dependency in the application manifest and
+will not be deployable without a min version being present. For Win32 programs and installers the
+Project Reunion bootstrapper will try to load the framework and error out with a similar
+instruction on how to resolve the missing dependencies Project Reunion.
+
+Developers must not assume that the Project Reunion main package is present, or that the version
+present is the minimum version required to support the feature set they expect. Instead, this API
+should be used to validate any Project Reunion requirements and direct the user to resolve any
+missing dependencies.
+
+The developer scenarios for this API is to answer the questions:
+1. Does this system have the min version of Project Reunion I require?
+2. What do I do to resolve this if it doesn't?
+
+The API addresses #1 by a simple query of the version they seek and a possible check for OS
+updates. #2 has two possible answers. First is direct resolution by the developer by acquiring and
+installing the packages if possible. Some programs will not have permission or motivation to do
+this, so the second answer is providing a link up to the user where the user can take action to
+install the package. The API will address this scenario by providing an answer as to what is
+required, where to get it, and where to go for instruction.
+
+# Examples
+
+## GetStatus
+
+This is the means by which an app or program can call to verify Project Reunion is present to a
+minimum version. It should be called once after activation before trying to use Project Reunion
+functionality. This method returns a ProjectReunionStatus object that describes the current
+deployment state of Project Reunion.
+
+```C++
+Windows::ApplicationModel::PackageVersion version{1, 0, 0, 0};
+auto status =
+Microsoft::Management::Deployment::ProjectReunionDeployment::GetStatus(version);
```

Comment:
I assume the "ProjectReunionDeployment" class name is temporary working name, and "ProjectReunion" will be dropped from these APIs?

Reply:
What installer? How does it "do this"?

An app could do none of this in main() and rely on its installer to DoTheRightThing(TM) and main() just assumes the system 'environmental soup' is correct. In which case this sample is what the installer would do.

But...why do you assume there's an installer? 'XCOPY-install' apps don't have an installer. And some devs want their app to 'check their environmental soup' (up front or if/when things go wrong).

NSIS or ClickOnce could do this, if they so choose. But apps can too, and some very much want or need to.

Reply:
Name is working and open to better suggestions. The reason this exists was defined at the top of the spec, but we may have OS updates required, we may have package updates required. We may not have the package at all. The developer needs to know and should be calling this API to check that all their expected Project Reunion dependencies are met and at a version minimum that they expect, otherwise their app will not function properly. So this API really has two functions:

1) Is Project Reunion the correct version and in a good state for what I need?  They call GetStatus() for this and provide the minimum version of Reunion they tested against.

2) If the status object returned is !IsOK(), what do I do about it? The obvious answer is their program should probably terminate and not try to continue without the dependency being satisfied, but the real question is how do they satisfy the dependency? There are two sub-options here for developers to resolve the problem.

a) They get the install instructions link and shell execute it and inform the user that there is action they need to take. While not great, this should be a 1-time deal for most situations. Sort of like needing to update DirectX or any other redist, but OS updates may be required here.

b) If OS updates are not required, and the are either a non-packaged medium-IL program, or a packaged app with Package Management capability, they can attempt to resolve the problem directly by getting the package (really appinstaller) install link and triggering the package install/update directly themselves.